### PR TITLE
Weighted objective

### LINKED ISSUE #24333: Allow multiple variables to contribute to an objective function
## Reason
I need an objective function that measures the misfit between multiple variables, like a displacement vector with disp_x and disp_y or temperature and displacement  

## Design
Use a weight to scale each variable sampled at every measurement point and use that scaled variable value when computing the objective function.  If measurements contained displacements in only the x and y direction, then the weights would be 1 or 0.  If the displacement measurement contains both x and y components, the weights would be the the vector components of that.

### PR DESCRIPTION
**Pull request by lynnmunday**
closes #24333

### CODE REVIEW

**Comment by lynnmunday** [comment]
@zachmprince will you review when you get a chance.

File: modules/optimization/include/dirackernels/ReporterTimePointSource.h [hunk 1]
```diff
@@ -10,37 +10,23 @@
...
-  /// y-coordinates from reporter
-  const std::vector<Real> & _coordy;
-  /// z-coordinates from reporter
-  const std::vector<Real> & _coordz;
   /// time-coordinates from reporter
   const std::vector<Real> & _coordt;
-  /// values from reporter
-  const std::vector<Real> & _values;
+
+private:
```
Comment (zachmprince) [inline modules/optimization/include/dirackernels/ReporterTimePointSource.h:28]: Why did you add this?

**Review by zachmprince — CHANGES_REQUESTED** [review]

File: modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/gold/main_out_optInfo_0001.csv [hunk 2]
```diff
@@ -1,3 +1,3 @@
 current_iterate,function_solves,function_value,gnorm,gradient_iterate,hessian_iterate,objective_iterate
-0,2,2347.5,15.738983573441,1,0,1
-1,8,0.0039592674042383,2.1870050037757e-12,1,3,1
+0,2,58409.624999999,469.07304717569,1,0,1
+1,8,0.0087767740012551,1.3865802329586e-06,1,3,1
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/gold/main_out_optInfo_0001.csv:3]: I am testing the weighting of the measurement variables.  I changed the weighting of each measurement point in measurmentData.csv so the second point is weighted 10x, and the third measurement is weighted by 1/10.  This changed the objective and gradient values computed.

File: modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i [hunk 3]
```diff
@@ -25,21 +25,22 @@
...
+  file_value = 'weightedMeasurement'
+  file_variable_weights = 'weight'
   # contents of measurementData.csv
   # measurement_points = '0.2 0.2 0
   #                       0.8 0.6 0
   #                       0.2 1.4 0
   #                       0.8 1.8 0'
   # measurement_values = '207 204 185 125'
-  outputs=none
+  # outputs=none
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i:36]: I'll delete

File: modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i [hunk 4]
```diff
@@ -25,21 +25,22 @@
...
 []
 
 [Executioner]
   type = Optimize
   tao_solver = taonls
   petsc_options_iname = '-tao_gttol -tao_nls_pc_type -tao_nls_ksp_type'
-  petsc_options_value = '1e-5 none cg'
+  petsc_options_value = '1e-3 none cg'
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i:43]: The weights made the convergence worst.  I weighted the second variable 10x more and the third 1/10x.  This lowered the overall convergence.

File: modules/optimization/test/tests/optimizationreporter/optimizationdata/tests [hunk 5]
```diff
@@ -0,0 +1,15 @@
+[Tests]
+  issues = '#24333'
+  design = 'OptimizationData.md'
+  [readData]
+    type = CSVDiff
+    max_threads = 1 # Optimize executioner does not support multiple threads
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/optimizationdata/tests:6]: It is not, it is only testing the optimimizationData reporter to see if it can sample two variables correctly and then correctly scale the simulation_values by the weights

File: modules/optimization/include/dirackernels/ReporterTimePointSource.h (hunk 1 above)
Comment (lynnmunday) [reply modules/optimization/include/dirackernels/ReporterTimePointSource.h:28]: I didn't add the private, you already had it private.

File: modules/optimization/src/dirackernels/ReporterTimePointSource.C [hunk 6]
```diff
@@ -94,17 +60,8 @@ ReporterTimePointSource::addPoints()
   for (const auto & i : make_range(nval))
     if (_coordt.empty() || MooseUtils::absoluteFuzzyEqual(at, _coordt[i]))
     {
-      Point pt;
-      pt(0) = _coordx.empty() ? 0.0 : _coordx[i];
-      pt(1) = _coordy.empty() ? 0.0 : _coordy[i];
-      pt(2) = _coordz.empty() ? 0.0 : _coordz[i];
+      Point pt(_coordx[i], _coordy[i], _coordz[i]);
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/dirackernels/ReporterTimePointSource.C:63]: I fixed this but we can't test it until [this pr](https://github.com/idaholab/moose/pull/23737) is merged to enable reporters to hold points.

File: modules/optimization/test/tests/optimizationreporter/optimizationdata/tests [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    rel_err = 0.00001
+    input = two_vars_file.i
+    csvdiff = two_vars_file_out_measure_data_0001.csv
+    # steady solve
+    recover = false
+    requirement = "The system shall be able to weight read variable weights and use "
+                  "the weight to scale the misfit."
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/optimizationdata/tests:13]: fixed

File: modules/optimization/test/tests/optimizationreporter/optimizationdata/tests [hunk 8]
```diff
@@ -0,0 +1,15 @@
+[Tests]
+  issues = '#24333'
+  design = 'OptimizationData.md'
+  [readData]
+    type = CSVDiff
+    max_threads = 1 # Optimize executioner does not support multiple threads
+    rel_err = 0.00001
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/optimizationdata/tests]: I removed it.

File: modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/gold/main_out_OptimizationReporter_0001.csv [hunk 9]
```diff
@@ -1,5 +0,0 @@
-grad_wrt_p,measurement_values,measurement_xcoord,measurement_ycoord,measurement_zcoord,misfit_values,p1,p2,simulation_values,measurement_time
-0,207,0.2,0.2,0,0.11599999999999999,113.26514660501002,287.85864907637,207,0
-0,204,0.8,0.6,0,-0.17800000000000002,0.0,0.0,204,0
-0,185,0.2,1.4,0,0.08900000000000001,0.0,0.0,185,0
-0,125,0.8,1.8,0,-0.2,0.0,0.0,125,0
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/gold/main_out_OptimizationReporter_0001.csv:5]: I added this file to the csvdiff.  I'm using the other file to check that it always converges in a single iteration when using a hessian.  The tolerances are loos because I only want to diff the measurement/simulation values and bcs. Everything else should go to zero.

[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/gold/main_out_OptimizationReporter_0001.csv:5]: I added this file to the csvdiff.  I'm using the other file to check that it always converges in a single iteration when using a hessian.  The tolerances are loos because I only want to diff the measurement/simulation values and bcs. Everything else should go to zero.

**Comment by lynnmunday** [comment]
@zachmprince will you review again and merge if its good?

**Review by zachmprince — APPROVED** [review]
